clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,269 @@
+                        return;
+                    }
+                    if expr_is_simple_identity(arm.pat, arm.body) {
+                        found_identity_arm = true;
+                    } else if expr_diverges(cx, arm.body) && pat_allowed_for_else(cx, arm.pat) {
+                        found_diverging_arm = true;
+                    }
+                }
+                if !(found_identity_arm && found_diverging_arm) {
+                    return;
+                }
+                span_lint(cx, MANUAL_LET_ELSE, stmt.span, "this could be rewritten as `let else`");
+            },
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+fn expr_diverges(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
+    fn is_never(cx: &LateContext<'_>, expr: &'_ Expr<'_>) -> bool {
+        if let Some(ty) = cx.typeck_results().expr_ty_opt(expr) {
+            return ty.is_never();
+        }
+        false
+    }
+    let mut does_diverge = false;
+    expr_visitor_no_bodies(|ex| {
+        match ex.kind {
+            ExprKind::Continue(_) | ExprKind::Break(_, _) | ExprKind::Ret(_) => {
+                does_diverge = true;
+                false
+            },
+            ExprKind::Call(call, _) => {
+                if is_never(cx, ex) || is_never(cx, call) {
+                    does_diverge = true;
+                    return false;
+                }
+                true
+            },
+            ExprKind::MethodCall(..) => {
+                if is_never(cx, ex) {
+                    does_diverge = true;
+                    return false;
+                }
+                true
+            },
+            ExprKind::If(if_expr, if_then, if_else) => {
+                let else_diverges = if_else.map_or(false, |ex| expr_diverges(cx, ex));
+                let diverges = expr_diverges(cx, if_expr) || (else_diverges && expr_diverges(cx, if_then));
+                if diverges {
+                    does_diverge = true;
+                }
+                false
+            },
+            ExprKind::Match(match_expr, match_arms, _) => {
+                let diverges =
+                    expr_diverges(cx, match_expr) || match_arms.iter().all(|arm| expr_diverges(cx, arm.body));
+                if diverges {
+                    does_diverge = true;
+                }
+                false
+            },
+
+            // Don't continue into loops, as they are breakable,
+            // and we'd have to start checking labels.
+            ExprKind::Loop(..) => false,
+
+            // Default: descend
+            // TODO: support breakable blocks. For this we would need label support.
+            _ => true,
+        }
+    })
+    .visit_expr(expr);
+    does_diverge
+}
+
+/// Returns true if the two spans come from different macro sites,
+/// or one comes from an invocation and the other is not from a macro at all.
+fn from_different_macros(span_a: Span, span_b: Span) -> bool {
```

Comment:
I couldn't find any such function. I've tried both searching for the name you gave me, as well as a function that takes two spans as arguments: `rg "fn.*\(.*Span.*Span.*\)"  .`

If you can point me to a function to use, I can try it.

Reply:
Ah, now that you mention it, we removed this function, because it got reduced to just `span_a.ctxt() != span_b.ctxt()`. This should also be enough here.